Add support for strongly typed entitlements for usePlanshipCustomer hook

pwojnaro opened the pull request:
This PR, when merged, will add support for a strongly typed entitlements wrapper class when using `usePlanshipCustomer` hook. 

Example:

```
export class ClickerEntitlements extends EntitlementsBase {

  get subscriptionButtonClicks(): number {
    return this.entitlementsDict?.['subscription-button-clicks'].valueOf()
  }
  
  get buttonClicksPerMinute(): number {
    return this.entitlementsDict?.['button-clicks-per-minute'].valueOf()
  }
  
  get maxProjects(): number {
    return this.entitlementsDict?.['max-projects'].valueOf()
  }
  
  get premiumButton(): boolean {
    return this.entitlementsDict?.['premium-button'].valueOf()
  }
  
  get analyticsPanel(): boolean {
    return this.entitlementsDict?.['analytics-panel'].valueOf()
  }
  
  get projectTypes(): [] {
    return this.entitlementsDict?.['project-types'].valueOf()
  }
}


{ entitlements } = usePlanshipCustomer<ClickerEntitlements>(ClickerEntitlements)

// entitlements is of ClickerEntitlements type

```

pwojnaro committed:
Add support for strongly typed entitlements for usePlanshipCustomer hook

trentpeterson commented on README.md:
```diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
 
+### Typed entitlements with PlanshipCustomer context provider
+
+When consuming PlanshipCustomer context in a TypeScript project, entitlements can be defined as a class with getters for individual levers, and passed to the `usePlanshipCustomer` hook.
```
-> *the* PlanshipCustomer context

Remove comma after levers

trentpeterson commented on README.md:
```diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
+
 ## Using Planship providers in a Next.js app
 
 By default, all components in a Next.js app are React Server Components. React Server Components don't allow for consuming React context *directly*, so Planship context providers cannot by used *as is*. To address this, simply wrap them your own Client Component as outlined in [this official guide](https://vercel.com/guides/react-context-state-management-nextjs#rendering-third-party-context-provi ...
```
wrap them your own -> wrap them in your own

trentpeterson commented on README.md:
````diff
@@ -198,11 +233,11 @@ export default async function RootLayout({
 ```
 
-A complete working example of a Nuxt.js app integrated with Planship can be found at https://github.com/planship/planship-nextjs-demo
+This pattern can be explored in more detail in our Next.js app example app at https://github.com/planship/planship-nextjs-example
````
Next.js app example app -> Next.js example app

trentpeterson commented on src/usePlanshipCustomer.ts:
```diff
@@ -1,9 +1,26 @@
+}): ICustomerContext<TEntititlements>
+export function usePlanshipCustomer(entitlementsType?: { new (entitlementsDict: Entitlements): object }) {
+  const currentContext = useContext<IPlanshipCustomerContext>(context)
+  if (entitlementsType)
```
Not sure if we're enforcing braces around if logic...

pwojnaro replied:
No, we do not. I just added it to eslint.rc, but I don't believe I enforce it in any other project (not a part of the recommended ruleset, seems arbitrary).

trentpeterson replied:
Yeah, just thinking we should be consistent. No strong opinion but if the logic is on a separate line than the subsequent code it seems like braces help make it clearer.

trentpeterson commented on static/entitlements-vscode.gif:
Should consider running this through an app to shrink the size a bit. Nearly 400kb - could likely be sped up, frames removed, etc.

pwojnaro replied:
I got it down to 220k

trentpeterson commented on README.md:
````diff
@@ -132,18 +134,51 @@ export default function SubscriptionInfoComponent({ children }) {
+}
+```
+
+This is helpful in certain IDEs like VS Code where it unlocks a full autocomplete experience for pricing entitlements entitlements.
````
Copy the language from the Vue module.

pwojnaro committed:
Address PR feedback

pwojnaro committed:
Add new intro language consistent with other SDKs

Additional README tweaks

Remove subscriptions from context, code/file cleanup

README tweaks

Change version for publishing

trentpeterson commented on README.md:
```diff
@@ -148,7 +148,7 @@ export default function SubscriptionInfoComponent({ children }) {
 ### Typed entitlements with PlanshipCustomer context provider
 
-When consuming PlanshipCustomer context in a TypeScript project, entitlements can be defined as a class with getters for individual levers, and passed to the `usePlanshipCustomer` hook.
+When working with the entitlements dictionary returned by `usePlanshipCustomer` hook, it can be useful to wrap it in an object with getters for individual levers.
```
returned by `usePlanshipCustomer` -> returned by the `usePlanshipCustomer`

pwojnaro reviewed (commented):
Addressed additional feedback

trentpeterson reviewed (approved):
LGTM
